src/main/java/com/github/mizool/jediscache/JedisCachingProvider.java

```diff
@@ -0,0 +1,211 @@
+import java.util.HashMap;
+import java.util.Properties;
+import java.util.WeakHashMap;
+
+import javax.cache.CacheException;
+import javax.cache.CacheManager;
+import javax.cache.configuration.OptionalFeature;
+import javax.cache.spi.CachingProvider;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.kohsuke.MetaInfServices;
+
+import redis.clients.jedis.JedisPool;
+import redis.clients.jedis.JedisPoolConfig;
+
+@Slf4j
+@MetaInfServices(CachingProvider.class)
+public class JedisCachingProvider implements CachingProvider
+{
+    private final JedisPool jedisPool;
+    private WeakHashMap<ClassLoader, HashMap<URI, CacheManager>> cacheManagersByClassLoader;
+
+    public JedisCachingProvider()
+    {
+        String host = System.getProperty("jediscache.host", "127.0.0.1");
+        log.info("JedisCache active, connecting to {}", host);
+        this.jedisPool = new JedisPool(new JedisPoolConfig(), host);
+        this.cacheManagersByClassLoader = new WeakHashMap<>();
+    }
+
+    @Override
+    public synchronized CacheManager getCacheManager(URI uri, ClassLoader classLoader, Properties properties)
+    {
+        if (uri == null)
+        {
+            uri = getDefaultURI();
+        }
+        if (classLoader == null)
+        {
+            classLoader = getDefaultClassLoader();
+        }
+        if (properties == null)
+        {
+            properties = new Properties();
+        }
+
+        HashMap<URI, CacheManager> cacheManagersByURI = cacheManagersByClassLoader.get(classLoader);
+
+        if (cacheManagersByURI == null)
+        {
+            cacheManagersByURI = new HashMap<>();
+        }
+
+        CacheManager cacheManager = cacheManagersByURI.get(uri);
+
+        if (cacheManager == null)
+        {
+            cacheManager = new JedisCacheManager(this, uri, classLoader, properties, jedisPool);
+            cacheManagersByURI.put(uri, cacheManager);
+        }
+
+        if (!cacheManagersByClassLoader.containsKey(classLoader))
+        {
+            cacheManagersByClassLoader.put(classLoader, cacheManagersByURI);
+        }
+
+        return cacheManager;
+    }
+
+    @Override
+    public CacheManager getCacheManager(URI uri, ClassLoader classLoader)
+    {
+        return getCacheManager(uri, classLoader, getDefaultProperties());
+    }
+
+    @Override
+    public CacheManager getCacheManager()
+    {
+        return getCacheManager(getDefaultURI(), getDefaultClassLoader(), null);
+    }
+
+    @Override
+    public ClassLoader getDefaultClassLoader()
+    {
+        return getClass().getClassLoader();
+    }
+
+    @Override
+    public URI getDefaultURI()
+    {
+        try
+        {
+            return new URI(this.getClass().getName());
+        }
+        catch (URISyntaxException e)
+        {
+            throw new CacheException("Failed to create the default URI for jedis cache", e);
+        }
+    }
+
+    @Override
+    public Properties getDefaultProperties()
+    {
+        return null;
```

Comment:
I wonder if this is allowed to return `null` or should return a `new Properties()` object. Sadly, the documentation is not specific about that. If we rewrite it to return a new properties object, it could even be used in `getCacheManager` -> `if (properties == null)` much like the `getDefaultClassLoader()` right above that. Thoughts?

Reply:
This code, along with like 90% of the code in this class, was copied&pasted from the reference implementation. So I assume that null is allowed, here.